cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -262,60 +296,43 @@ public ArtifactData apply(DatasetContext<Table> context) throws Exception {
+              ArtifactColumn artifactColumn = ArtifactColumn.parse(column.getKey());
+              PluginData pluginData = gson.fromJson(Bytes.toString(column.getValue()), PluginData.class);
+              // filter out plugins that don't extend this version of the parent artifact
+              if (matches(pluginData, parentArtifactId.getVersion())) {
+                ArtifactInfo artifactInfo = new ArtifactInfo(artifactColumn.artifactId, pluginData.artifactLocation);
+                result.put(artifactInfo, pluginData.pluginClass);
+              }
+            }
           }
           return result;
         }
       });
     if (plugins.isEmpty()) {
-      throw new PluginNotExistsException(namespace, type, name);
+      throw new PluginNotExistsException(parentArtifactId.getNamespace(), type, name);
     }
     return plugins;
   }
 
-  /**
-   * Get the plugin class for a specific plugin type, name, and belonging to a specific artifact.
-   * Results are returned as a map from artifact to plugins in that artifact.
-   *
-   * @param artifactId the id of the artifact the plugin is from
-   * @param type the type of plugin to look for
-   * @param name the name of the plugin to look for
-   * @return a map entry containing the artifact info and plugin class details. Never null
-   * @throws PluginNotExistsException if no such plugin exists
-   * @throws IOException if there was an exception reading metadata from the metastore
-   */
-  public Map.Entry<ArtifactInfo, PluginClass> getPluginClass(
-    final Id.Artifact artifactId, final String type, final String name) throws IOException, PluginNotExistsException {
-    Map.Entry<ArtifactInfo, PluginClass> plugin = metaTable.executeUnchecked(
-      new TransactionExecutor.Function<DatasetContext<Table>, Map.Entry<ArtifactInfo, PluginClass>>() {
-        @Override
-        public Map.Entry<ArtifactInfo, PluginClass> apply(DatasetContext<Table> context) throws Exception {
-          PluginKey pluginKey = new PluginKey(artifactId.getNamespace(), type, name);
-          ArtifactColumn column = new ArtifactColumn(artifactId.getVersion().getVersion(), artifactId.getName());
-          byte[] value = context.get().get(pluginKey.getRowKey(), column.getColumn());
-          if (value == null) {
-            return null;
-          }
-          PluginData pluginData = gson.fromJson(Bytes.toString(value), PluginData.class);
-          return Maps.immutableEntry(new ArtifactInfo(artifactId, pluginData.artifactLocation), pluginData.pluginClass);
-        }
-      });
-    if (plugin == null) {
-      throw new PluginNotExistsException(artifactId, type, name);
-    }
-    return plugin;
+  private boolean matches(PluginData pluginData, ArtifactVersion version) {
+    ArtifactVersion lower = new ArtifactVersion(pluginData.parentVersionLower);
+    ArtifactVersion upper = new ArtifactVersion(pluginData.parentVersionUpper);
+    return lower.compareTo(version) <= 0 && upper.compareTo(version) > 0;
```

Comment:
Would it be more readable if we switch the logic to do version.compareTo(lower) >=0 && version.compareTo(upper) < 0 ?